cpp/src/fil/treelite_import.cu

```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
   return 0;  // Treelite format assumes that the root is 0
 }
 
-template <typename T, typename L>
-inline int max_depth(const tl::Tree<T, L>& tree)
+struct empty {
+};
+
+/** walk a Treelite tree, visiting each inner node with inner_func and each leaf node with
+  leaf_func. See walk_tree::invitation::state documentation for how RootToLeafPathState is retained
+during traversal. Any per-tree state during traversal should be captured by the lambdas themselves.
+  inner_func(int node_id, RootToLeafPathState state) should return a pair of new states, one for
+each child node. leaf_func(int, RootToLeafPathState) returns nothing.
+**/
+template <typename RootToLeafPathState,
+          typename Threshold,
+          typename Leaf,
+          typename InnerNodeVisitFunc,
+          typename LeafNodeVisitFunc = decltype([](int, RootToLeafPathState) {})>
+inline void walk_tree(
+  const tl::Tree<Threshold, Leaf>& tree,
+  InnerNodeVisitFunc inner_func,
+  LeafNodeVisitFunc leaf_func = [](int, RootToLeafPathState) {})
 {
-  // trees of this depth aren't used, so it most likely means bad input data,
-  // e.g. cycles in the forest
-  const int DEPTH_LIMIT = 500;
-  int root_index        = tree_root(tree);
-  typedef std::pair<int, int> pair_t;
-  std::stack<pair_t> stack;
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
+  /// needed to visit a node
+  struct invitation {
+    int tl_node_id;
+    /// Retained while visiting nodes on a single path from root to leaf.
+    /// This generalizes the node index that's carried over during inference tree traversal.
+    RootToLeafPathState state;
+  };
+  std::stack<invitation> stack;
+  stack.push(invitation{tree_root(tree), RootToLeafPathState()});
   while (!stack.empty()) {
-    const pair_t& pair = stack.top();
-    int node_id        = pair.first;
-    int depth          = pair.second;
+    invitation i = stack.top();
     stack.pop();
-    while (!tree.IsLeaf(node_id)) {
-      stack.push(pair_t(tree.LeftChild(node_id), depth + 1));
-      node_id = tree.RightChild(node_id);
-      depth++;
-      ASSERT(depth < DEPTH_LIMIT, "depth limit reached, might be a cycle in the tree");
+    while (!tree.IsLeaf(i.tl_node_id)) {
+      auto [left_state, right_state] = inner_func(i.tl_node_id, i.state);
```

Comment:
This loop is certainly more concise, but at the price of having to return a pair of empty values if we don't care about the state.

I think there should be a separate wrapper (also named `walk_tree()`) for the case when the state is empty. It can then wrap `leaf_func` into something that returns a tuple of empty values and accepts an extra empty argument.